src/overlay021/pokedexfieldmap.c

```diff
@@ -0,0 +1,279 @@
+static void PopulateMapAtXY(u8 *pokedexFieldMap, int y, int x, int mapHeight, int mapWidth);
+static int XYToValue(u8 *pokedexFieldMap, int y, int x, int mapHeight, int mapWidth);
+static void FindNeighboringFields(u8 *pokedexFieldMap, int *neighbors, int y, int x, int mapHeight, int mapWidth);
+static int GetNeighborType(int neighbors);
+
+void PokedexFieldMap_DisplayFields(Window *window, u8 *src, u16 fieldWidth, u16 fieldHeight, u8 mapSize, u8 *pokedexFieldMap, u8 mapHeight, u8 mapWidth, u16 yOffset, u16 xOffset)
+{
+    for (int x = 0; x < mapWidth; x++) {
+        for (int y = 0; y < mapHeight; y++) {
+            DisplayOnWindow(window, src, fieldWidth, fieldHeight, mapSize, pokedexFieldMap[(x * mapWidth) + y], yOffset + (y * mapSize), xOffset + (x * mapSize));
+        }
+    }
+}
+
+void PokedexFieldMap_SmoothFields(u8 *pokedexFieldMap, u8 mapHeight, u8 mapWidth)
+{
+    for (int x = 0; x < mapWidth; x++) {
+        for (int y = 0; y < mapHeight; y++) {
+            PopulateMapAtXY(pokedexFieldMap, y, x, mapHeight, mapWidth);
+        }
+    }
+}
+
+static void DisplayOnWindow(Window *window, u8 *src, u16 fieldWidth, u16 fieldHeight, int mapSize, int fieldType, u8 y, u8 x)
+{
+    if (fieldType != 0) {
+        int srcX = fieldType * mapSize;
+        int srcY = srcX / fieldWidth;
+        srcX = srcX % fieldWidth;
+
+        Window_BlitBitmapRect(window, src, srcX, srcY, fieldWidth, fieldHeight, y, x, mapSize, mapSize);
+    }
+}
+
+static void PopulateMapAtXY(u8 *pokedexFieldMap, int y, int x, int mapHeight, int mapWidth)
+{
+    int neighbors;
+    int fieldType;
+
+    if (pokedexFieldMap[(x * mapHeight) + y] == 0) {
+        FindNeighboringFields(pokedexFieldMap, &neighbors, y, x, mapHeight, mapWidth);
+
+        fieldType = GetNeighborType(neighbors);
+        pokedexFieldMap[(x * mapHeight) + y] = fieldType;
+    }
+}
+
+static int XYToValue(u8 *pokedexFieldMap, int y, int x, int mapHeight, int mapWidth)
+{
+    if ((y < 0) || (y >= mapHeight)) {
+        return -1;
+    }
+
+    if ((x < 0) || (x >= mapWidth)) {
+        return -1;
+    }
+
+    return pokedexFieldMap[(x * mapHeight) + y];
+}
+
+static void FindNeighboringFields(u8 *pokedexFieldMap, int *neighbors, int y, int x, int mapHeight, int mapWidth)
+{
+    *neighbors = FN_ALONE; // binary storage of fields (NE)(SE)(NW)(SW)(N)(S)(E)(W)
+    // only records intercardinal directions if adjacent cardinals are both empty
+
+    if (XYToValue(pokedexFieldMap, y, x - 1, mapHeight, mapWidth) == 1) {
+        *neighbors |= FN_WEST;
+    }
+
+    if (XYToValue(pokedexFieldMap, y, x + 1, mapHeight, mapWidth) == 1) {
+        *neighbors |= FN_EAST;
+    }
+
+    if (XYToValue(pokedexFieldMap, y - 1, x, mapHeight, mapWidth) == 1) {
+        *neighbors |= FN_SOUTH;
+    }
+
+    if (XYToValue(pokedexFieldMap, y + 1, x, mapHeight, mapWidth) == 1) {
+        *neighbors |= FN_NORTH;
+    }
+
+    if ((((*neighbors) & FN_WEST) == 0) && (((*neighbors) & FN_SOUTH) == 0)) {
```

Comment:
No need to nest your parens so deep like this. We usually prefer as few as possible while still maintaining clarity, e.g.,

```c
if ((*neighbors & FN_WEST) == 0 && (*neighbors & FN_SOUTH) == 0) {
```